### PR TITLE
riscv64: Add Zfh extension support

### PR DESCRIPTION
**Pull request by rjiejie**
This patch make following simple design rules:
- Construct common encoding interface/macro for frond-end & back-end, see VEX/pub/libvex_riscv_common.h
- Every Independent ISA extension is modular and splited as individual file, that will help to maintain all extensions of RISCV

### CODE REVIEW

**Review by petrpavlu — CHANGES_REQUESTED** [review]
Thanks for this work. I've added initial review comments but still need to go through some parts.

One general problem with this new code to support Zfh is that it is largely inconsistent with the current RISC-V port. It goes in several aspects on its own way where I don't see justification for doing that. Some differences look to me purely based on subjective matter of style.

The current RISC-V port tries to be fairly consistent with the upstream code, particularly with the AArch64 and MIPS variants. There is usually very little reason to deviate from them, both in style and in actual logic. The RISC-V port has only a limited set of things where I decided to do something slightly differently, usually for good reasons.

For instance, as already touched upon in https://github.com/petrpavlu/valgrind-riscv64/issues/3#issuecomment-1362164866, I think it is ok that `VEX/priv/guest_riscv64_toIR.c` is only one file and not split per extensions. The file is reasonably small and the frontend code is very linear, it simply decodes instructions one by one. It doesn't look to me that having Zfh in a separate file `VEX/priv/guest_riscv64Zfh_toIR.c` makes the code easier to navigate or maintain. My view is that it only creates an inconsistency from the rest of the VEX code for no actual benefit.

My second general comment  is about adding new Iops for FP16. That looks fine to me as they should be eventually needed also for remaining FP16 support on AArch64.

However, it makes me wonder how to best approach merging this work. My thinking is to initially submit upstream for review only the current RV64GC support, make maintainers aware that some new Iops will be needed for RISC-V, and once the port lands, have a separate upstream review for the FP16 support.

That also means that I would merge this FP16 work in this repository on a separate branch for now, or the other way around, create a new branch for only the code that should go upstream in the first batch.

File: coregrind/m_machine.c
Comment (paulfloyd) [inline coregrind/m_machine.c:2276]: #if defined is more consistent with the existing code (rather than #ifdef)

File: VEX/pub/libvex_basictypes.h
Comment (paulfloyd) [inline VEX/pub/libvex_basictypes.h:94]: _Float16 isn't RISCV specific. I'm not sure which version of C added it. From what I see there is no straightforward feature test for this. See https://stackoverflow.com/questions/69977196/how-to-correctly-determine-at-compile-time-that-float16-is-supported

It might be worth adding a configure time check for this.

**Comment by rjiejie** [comment]
> However, it makes me wonder how to best approach merging this work. My thinking is to initially submit upstream for review only the current RV64GC support, make maintainers aware that some new Iops will be needed for RISC-V, and once the port lands, have a separate upstream review for the FP16 support.
> 
> That also means that I would merge this FP16 work in this repository on a separate branch for now, or the other way around, create a new branch for only the code that should go upstream in the first batch.

Thanks for all comments @petrpavlu @paulfloyd 
I agree with your opinion, we should upstream RVGC as base code for RISCV first :)

**Pull request closed without merging**
